Storage Drivers (starting w/ subnet memory)

jzelinskie opened the pull request:
This is a WIP branch attempting to fix #267.

mrd0ll4r commented on cmd/chihaya/config.go:
```diff
@@ -63,6 +72,16 @@ func ParseConfigFile(path string) (*ConfigFile, error) {
+// NewPeerStore allocates a storage.PeerStore for a given ConfigFile.
+func (cfg ConfigFile) NewPeerStore() (storage.PeerStore, error) {
+	storecfg := cfg.MainConfigBlock.Storage
+	ps, exists, err := storage.NewPeerStore(storecfg.Name, storecfg.Config)
```
Huh, I would've expected something like an `ErrUnknownDriver`, but this works too. Try something new from time to time :)

jzelinskie replied:
I've seen this pattern used in a few other projects. It'd simplify things to be an error, but less explicit. I'm open to both, tbh.

mrd0ll4r replied:
Yeah let's do this, looks good :)

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,595 @@
+	PeerLifetime                time.Duration `yaml:"peer_lifetime"`
+	ShardCount                  int           `yaml:"shard_count"`
+	PreferredIPv4SubnetMaskBits int           `yaml:"preferred_ipv4_subnet_mask_bits"`
+	PreferredIPv6SubnetMaskBits int           `yaml:"preferred_ipv6_subnet_mask_bits"`
```
This means the bit count rather than the actual bits, right? If yes: name is misleading.

jzelinskie replied:
Agreed. Any idea for a good name for this that isn't too long?

mrd0ll4r replied:
`(Preferred)IPv4SubnetLength` maybe?

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,595 @@
+}
+
+type swarm struct {
+	// map serialized peer to mtime
```
This should be updated to describe what is really happening here. I remember when I first read the v1 sources and didn't really grasp the idea :)

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,595 @@
+
+func (s swarm) lenSeeders() (i int) {
+	for _, subnet := range s.seeders {
+		for range subnet {
```
Why are you not using `len(subnet)` here?

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,595 @@
+
+func (s swarm) lenLeechers() (i int) {
+	for _, subnet := range s.leechers {
+		for range subnet {
```
Same as above

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,595 @@
+	default:
+		panic("IP is neither v4 nor v6")
+	}
+	return maskedIP.String()
```
We could experiment with just using the masked IP printed in hexadecimal, might be faster than using the "real" string representation.

jzelinskie replied:
Masks are slices of bytes, which means they're dynamically allocated and can't be used as keys in maps 😿

mrd0ll4r replied:
Ignore my previous comment, `String` does just what I asked (hex print the bytes)...
We could maybe just use `string(maskedIP)` though?

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,595 @@
+		recordInfohashesDelta(1)
+	}
+
+	mask := s.mask(p)
```
I'd call this `subnet` to make it clearer what's going on.

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,595 @@
+	}
+
+	mask := s.mask(p)
+	delete(shard.swarms[ih].leechers[mask], pk)
```
`shard.swarms[ih].leechers[mask]` can be nil if we just created `shard.swarms[ih]` above. Does `delete` panic if called on a nil map?

jzelinskie replied:
nope

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,595 @@
+			}
+
+			for subnet := range shard.swarms[ih].seeders {
+				for pk, mtime := range shard.swarms[ih].leechers[subnet] {
```
Copypasta, should be `shard.swarms[ih].seeders[subnet]`

jzelinskie replied:
nice catch

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,595 @@
+				}
+			}
+
+			// TODO(jzelinskie): fix this to sum all peers in all subnets
```
This TODO looks like it should be fixed because it could delete peers?

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,595 @@
+					if mtime <= cutoffUnix {
+						delete(shard.swarms[ih].seeders[subnet], pk)
+					}
+				}
```
Delete subnet if empty?

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,595 @@
+					if mtime <= cutoffUnix {
+						delete(shard.swarms[ih].leechers[subnet], pk)
+					}
+				}
```
Delete subnet if empty?

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,595 @@
+type serializedPeer string
+
+type peerShard struct {
+	swarms map[bittorrent.InfoHash]swarm
```
We should make some assumptions about the number of infohashes, peers and subnets. Based on that we might want to have a `map[subnet]map[InfoHash]swarm` here to save the amount of maps (because they use a lot of memory).

jzelinskie commented on storage/storage.go:
```diff
@@ -1,10 +1,22 @@
 
+var (
+	driversM sync.RWMutex
+	drivers  = make(map[string]Driver)
```
Doesn't really buy you anything. Let's keep it simple.

mrd0ll4r commented:
Oh I forgot: Also needs docs for the config and something that describes how peers are selected

jzelinskie commented:
Rebased and addressed a bunch of comments and ran benchmarks:

```
BenchmarkPut-2                                   3000000               465 ns/op
BenchmarkPut1k-2                                 3000000               553 ns/op
BenchmarkPut1kInfohash-2                         5000000               386 ns/op
BenchmarkPut1kInfohash1k-2                       5000000               362 ns/op
BenchmarkPutDelete-2                             1000000              1552 ns/op
BenchmarkPutDelete1k-2                           1000000              1571 ns/op
BenchmarkPutDelete1kInfohash-2                   1000000              1682 ns/op
BenchmarkPutDelete1kInfohash1k-2                 1000000              1672 ns/op
BenchmarkDeleteNonexist-2                       10000000               120 ns/op
BenchmarkDeleteNonexist1k-2                     10000000               113 ns/op
BenchmarkDeleteNonexist1kInfohash-2             20000000                96.8 ns/op
BenchmarkDeleteNonexist1kInfohash1k-2           20000000               100 ns/op
BenchmarkPutGradDelete-2                          500000              2516 ns/op
BenchmarkPutGradDelete1k-2                        500000              2683 ns/op
BenchmarkPutGradDelete1kInfohash-2                500000              2685 ns/op
BenchmarkPutGradDelete1kInfohash1k-2              500000              2866 ns/op
BenchmarkGradNonexist-2                          3000000               510 ns/op
BenchmarkGradNonexist1k-2                        2000000               583 ns/op
BenchmarkGradNonexist1kInfohash-2                5000000               365 ns/op
BenchmarkGradNonexist1kInfohash1k-2              5000000               384 ns/op
BenchmarkAnnounceLeecher-2                         50000             44568 ns/op
BenchmarkAnnounceLeecher1kInfohash-2               20000             86686 ns/op
BenchmarkAnnounceSeeder-2                          50000             40797 ns/op
BenchmarkAnnounceSeeder1kInfohash-2                20000             91576 ns/op
PASS
ok      github.com/chihaya/chihaya/storage/memorybysubnet       68.414s

BenchmarkPut-2                                  10000000               232 ns/op
BenchmarkPut1k-2                                 5000000               256 ns/op
BenchmarkPut1kInfohash-2                        10000000               177 ns/op
BenchmarkPut1kInfohash1k-2                      10000000               182 ns/op
BenchmarkPutDelete-2                             2000000               816 ns/op
BenchmarkPutDelete1k-2                           2000000               754 ns/op
BenchmarkPutDelete1kInfohash-2                   2000000               802 ns/op
BenchmarkPutDelete1kInfohash1k-2                 2000000               794 ns/op
BenchmarkDeleteNonexist-2                       10000000               119 ns/op
BenchmarkDeleteNonexist1k-2                     10000000               112 ns/op
BenchmarkDeleteNonexist1kInfohash-2             20000000                86.6 ns/op
BenchmarkDeleteNonexist1kInfohash1k-2           20000000                92.5 ns/op
BenchmarkPutGradDelete-2                         1000000              1167 ns/op
BenchmarkPutGradDelete1k-2                       1000000              1167 ns/op
BenchmarkPutGradDelete1kInfohash-2               1000000              1237 ns/op
BenchmarkPutGradDelete1kInfohash1k-2             1000000              1232 ns/op
BenchmarkGradNonexist-2                          5000000               261 ns/op
BenchmarkGradNonexist1k-2                        5000000               269 ns/op
BenchmarkGradNonexist1kInfohash-2               10000000               200 ns/op
BenchmarkGradNonexist1kInfohash1k-2             10000000               197 ns/op
BenchmarkAnnounceLeecher-2                        200000              6028 ns/op
BenchmarkAnnounceLeecher1kInfohash-2              200000              8268 ns/op
BenchmarkAnnounceSeeder-2                         200000              6095 ns/op
BenchmarkAnnounceSeeder1kInfohash-2               200000              7750 ns/op
PASS
ok      github.com/chihaya/chihaya/storage/memory       49.556s
```

jzelinskie commented:
Rebased and addressed a bunch of comments and ran benchmarks:

```
BenchmarkPut-2                                   3000000               465 ns/op
BenchmarkPut1k-2                                 3000000               553 ns/op
BenchmarkPut1kInfohash-2                         5000000               386 ns/op
BenchmarkPut1kInfohash1k-2                       5000000               362 ns/op
BenchmarkPutDelete-2                             1000000              1552 ns/op
BenchmarkPutDelete1k-2                           1000000              1571 ns/op
BenchmarkPutDelete1kInfohash-2                   1000000              1682 ns/op
BenchmarkPutDelete1kInfohash1k-2                 1000000              1672 ns/op
BenchmarkDeleteNonexist-2                       10000000               120 ns/op
BenchmarkDeleteNonexist1k-2                     10000000               113 ns/op
BenchmarkDeleteNonexist1kInfohash-2             20000000                96.8 ns/op
BenchmarkDeleteNonexist1kInfohash1k-2           20000000               100 ns/op
BenchmarkPutGradDelete-2                          500000              2516 ns/op
BenchmarkPutGradDelete1k-2                        500000              2683 ns/op
BenchmarkPutGradDelete1kInfohash-2                500000              2685 ns/op
BenchmarkPutGradDelete1kInfohash1k-2              500000              2866 ns/op
BenchmarkGradNonexist-2                          3000000               510 ns/op
BenchmarkGradNonexist1k-2                        2000000               583 ns/op
BenchmarkGradNonexist1kInfohash-2                5000000               365 ns/op
BenchmarkGradNonexist1kInfohash1k-2              5000000               384 ns/op
BenchmarkAnnounceLeecher-2                         50000             44568 ns/op
BenchmarkAnnounceLeecher1kInfohash-2               20000             86686 ns/op
BenchmarkAnnounceSeeder-2                          50000             40797 ns/op
BenchmarkAnnounceSeeder1kInfohash-2                20000             91576 ns/op
PASS
ok      github.com/chihaya/chihaya/storage/memorybysubnet       68.414s

BenchmarkPut-2                                  10000000               232 ns/op
BenchmarkPut1k-2                                 5000000               256 ns/op
BenchmarkPut1kInfohash-2                        10000000               177 ns/op
BenchmarkPut1kInfohash1k-2                      10000000               182 ns/op
BenchmarkPutDelete-2                             2000000               816 ns/op
BenchmarkPutDelete1k-2                           2000000               754 ns/op
BenchmarkPutDelete1kInfohash-2                   2000000               802 ns/op
BenchmarkPutDelete1kInfohash1k-2                 2000000               794 ns/op
BenchmarkDeleteNonexist-2                       10000000               119 ns/op
BenchmarkDeleteNonexist1k-2                     10000000               112 ns/op
BenchmarkDeleteNonexist1kInfohash-2             20000000                86.6 ns/op
BenchmarkDeleteNonexist1kInfohash1k-2           20000000                92.5 ns/op
BenchmarkPutGradDelete-2                         1000000              1167 ns/op
BenchmarkPutGradDelete1k-2                       1000000              1167 ns/op
BenchmarkPutGradDelete1kInfohash-2               1000000              1237 ns/op
BenchmarkPutGradDelete1kInfohash1k-2             1000000              1232 ns/op
BenchmarkGradNonexist-2                          5000000               261 ns/op
BenchmarkGradNonexist1k-2                        5000000               269 ns/op
BenchmarkGradNonexist1kInfohash-2               10000000               200 ns/op
BenchmarkGradNonexist1kInfohash1k-2             10000000               197 ns/op
BenchmarkAnnounceLeecher-2                        200000              6028 ns/op
BenchmarkAnnounceLeecher1kInfohash-2              200000              8268 ns/op
BenchmarkAnnounceSeeder-2                         200000              6095 ns/op
BenchmarkAnnounceSeeder1kInfohash-2               200000              7750 ns/op
PASS
ok      github.com/chihaya/chihaya/storage/memory       49.556s
```

jzelinskie commented:
Rebased and addressed a bunch of comments and ran benchmarks:

```
BenchmarkPut-2                                   3000000               465 ns/op
BenchmarkPut1k-2                                 3000000               553 ns/op
BenchmarkPut1kInfohash-2                         5000000               386 ns/op
BenchmarkPut1kInfohash1k-2                       5000000               362 ns/op
BenchmarkPutDelete-2                             1000000              1552 ns/op
BenchmarkPutDelete1k-2                           1000000              1571 ns/op
BenchmarkPutDelete1kInfohash-2                   1000000              1682 ns/op
BenchmarkPutDelete1kInfohash1k-2                 1000000              1672 ns/op
BenchmarkDeleteNonexist-2                       10000000               120 ns/op
BenchmarkDeleteNonexist1k-2                     10000000               113 ns/op
BenchmarkDeleteNonexist1kInfohash-2             20000000                96.8 ns/op
BenchmarkDeleteNonexist1kInfohash1k-2           20000000               100 ns/op
BenchmarkPutGradDelete-2                          500000              2516 ns/op
BenchmarkPutGradDelete1k-2                        500000              2683 ns/op
BenchmarkPutGradDelete1kInfohash-2                500000              2685 ns/op
BenchmarkPutGradDelete1kInfohash1k-2              500000              2866 ns/op
BenchmarkGradNonexist-2                          3000000               510 ns/op
BenchmarkGradNonexist1k-2                        2000000               583 ns/op
BenchmarkGradNonexist1kInfohash-2                5000000               365 ns/op
BenchmarkGradNonexist1kInfohash1k-2              5000000               384 ns/op
BenchmarkAnnounceLeecher-2                         50000             44568 ns/op
BenchmarkAnnounceLeecher1kInfohash-2               20000             86686 ns/op
BenchmarkAnnounceSeeder-2                          50000             40797 ns/op
BenchmarkAnnounceSeeder1kInfohash-2                20000             91576 ns/op
PASS
ok      github.com/chihaya/chihaya/storage/memorybysubnet       68.414s

BenchmarkPut-2                                  10000000               232 ns/op
BenchmarkPut1k-2                                 5000000               256 ns/op
BenchmarkPut1kInfohash-2                        10000000               177 ns/op
BenchmarkPut1kInfohash1k-2                      10000000               182 ns/op
BenchmarkPutDelete-2                             2000000               816 ns/op
BenchmarkPutDelete1k-2                           2000000               754 ns/op
BenchmarkPutDelete1kInfohash-2                   2000000               802 ns/op
BenchmarkPutDelete1kInfohash1k-2                 2000000               794 ns/op
BenchmarkDeleteNonexist-2                       10000000               119 ns/op
BenchmarkDeleteNonexist1k-2                     10000000               112 ns/op
BenchmarkDeleteNonexist1kInfohash-2             20000000                86.6 ns/op
BenchmarkDeleteNonexist1kInfohash1k-2           20000000                92.5 ns/op
BenchmarkPutGradDelete-2                         1000000              1167 ns/op
BenchmarkPutGradDelete1k-2                       1000000              1167 ns/op
BenchmarkPutGradDelete1kInfohash-2               1000000              1237 ns/op
BenchmarkPutGradDelete1kInfohash1k-2             1000000              1232 ns/op
BenchmarkGradNonexist-2                          5000000               261 ns/op
BenchmarkGradNonexist1k-2                        5000000               269 ns/op
BenchmarkGradNonexist1kInfohash-2               10000000               200 ns/op
BenchmarkGradNonexist1kInfohash1k-2             10000000               197 ns/op
BenchmarkAnnounceLeecher-2                        200000              6028 ns/op
BenchmarkAnnounceLeecher1kInfohash-2              200000              8268 ns/op
BenchmarkAnnounceSeeder-2                         200000              6095 ns/op
BenchmarkAnnounceSeeder1kInfohash-2               200000              7750 ns/op
PASS
ok      github.com/chihaya/chihaya/storage/memory       49.556s
```

mrd0ll4r commented:
Benchcmp:

```
benchmark                                 old ns/op     new ns/op     delta
BenchmarkPut-2                            232           465           +100.43%
BenchmarkPut1k-2                          256           553           +116.02%
BenchmarkPut1kInfohash-2                  177           386           +118.08%
BenchmarkPut1kInfohash1k-2                182           362           +98.90%
BenchmarkPutDelete-2                      816           1552          +90.20%
BenchmarkPutDelete1k-2                    754           1571          +108.36%
BenchmarkPutDelete1kInfohash-2            802           1682          +109.73%
BenchmarkPutDelete1kInfohash1k-2          794           1672          +110.58%
BenchmarkDeleteNonexist-2                 119           120           +0.84%
BenchmarkDeleteNonexist1k-2               112           113           +0.89%
BenchmarkDeleteNonexist1kInfohash-2       86.6          96.8          +11.78%
BenchmarkDeleteNonexist1kInfohash1k-2     92.5          100           +8.11%
BenchmarkPutGradDelete-2                  1167          2516          +115.60%
BenchmarkPutGradDelete1k-2                1167          2683          +129.91%
BenchmarkPutGradDelete1kInfohash-2        1237          2685          +117.06%
BenchmarkPutGradDelete1kInfohash1k-2      1232          2866          +132.63%
BenchmarkGradNonexist-2                   261           510           +95.40%
BenchmarkGradNonexist1k-2                 269           583           +116.73%
BenchmarkGradNonexist1kInfohash-2         200           365           +82.50%
BenchmarkGradNonexist1kInfohash1k-2       197           384           +94.92%
BenchmarkAnnounceLeecher-2                6028          44568         +639.35%
BenchmarkAnnounceLeecher1kInfohash-2      8268          86686         +948.45%
BenchmarkAnnounceSeeder-2                 6095          40797         +569.35%
BenchmarkAnnounceSeeder1kInfohash-2       7750          91576         +1081.63%
```

jzelinskie commented:
It's probably slow because of how many new maps it has to allocate to run the tests.

I had a realization while working on rebasing this the other day: map deletes never shrink the map, so the GC actually needs to allocate a new map and copy things over every now and then to actually free memory.

mrd0ll4r commented:
> I had a realization while working on rebasing this the other day: map deletes never shrink the map, so the GC actually needs to allocate a new map and copy things over every now and then to actually free memory.

Wait, what?

On the PR: I'm gonna approve it if you add some documentation on how it works/how peers are selected/how it's configured.

jzelinskie commented:
@mrd0ll4r this is ready for review now.

>> I had a realization while working on rebasing this the other day: map deletes never shrink the map, so the GC actually needs to allocate a new map and copy things over every now and then to actually free memory.
>
>Wait, what?

https://github.com/golang/go/issues/20135

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,780 @@
+func (ps *peerStore) PutSeeder(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped memory store")
```
Should we change this to `memorysubnet` ? Or maybe just leave out the name altogether?

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,780 @@
+			seeders:  make(map[peerSubnet]map[serializedPeer]int64),
+			leechers: make(map[peerSubnet]map[serializedPeer]int64),
+		}
+		recordInfohashesDelta(1)
```
No need

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,780 @@
+	}
+
+	// If this peer isn't already a seeder, update the stats for the swarm.
+	if _, ok := shard.swarms[ih].seeders[preferredSubnet]; !ok {
```
This should be `shard.swarms[ih].seeders[preferredSubnet][pk]`

jzelinskie replied:
nice save

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,780 @@
+	}
+
+	preferredSubnet := newPeerSubnet(p.IP, ps.ipv4Mask, ps.ipv6Mask)
+	if _, ok := shard.swarms[ih].seeders[preferredSubnet][pk]; !ok {
```
This could panic if `shard.swarms[ih].seeders[preferredSubnet] == nil`, right? We have to check for that first.

jzelinskie replied:
Nope: https://play.golang.org/p/6IR3kPBN3R

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,780 @@
+
+	if shard.swarms[ih].lenSeeders()|shard.swarms[ih].lenLeechers() == 0 {
+		delete(shard.swarms, ih)
+		recordInfohashesDelta(-1)
```
No need

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,780 @@
+			seeders:  make(map[peerSubnet]map[serializedPeer]int64),
+			leechers: make(map[peerSubnet]map[serializedPeer]int64),
+		}
+		recordInfohashesDelta(1)
```
No need

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,780 @@
+	}
+
+	preferredSubnet := newPeerSubnet(p.IP, ps.ipv4Mask, ps.ipv6Mask)
+	if _, ok := shard.swarms[ih].leechers[preferredSubnet][pk]; !ok {
```
Same as with `DeleteSeeder`

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,780 @@
+
+	if shard.swarms[ih].lenSeeders()|shard.swarms[ih].lenLeechers() == 0 {
+		delete(shard.swarms, ih)
+		recordInfohashesDelta(-1)
```
No need

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,780 @@
+			seeders:  make(map[peerSubnet]map[serializedPeer]int64),
+			leechers: make(map[peerSubnet]map[serializedPeer]int64),
+		}
+		recordInfohashesDelta(1)
```
No need

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,780 @@
+
+	// If this peer is a leecher, update the stats for the swarm and remove them.
+	preferredSubnet := newPeerSubnet(p.IP, ps.ipv4Mask, ps.ipv6Mask)
+	if _, ok := shard.swarms[ih].leechers[preferredSubnet][pk]; ok {
```
Check for subnet first

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,780 @@
+	}
+
+	// If this peer isn't already a seeder, update the stats for the swarm.
+	if _, ok := shard.swarms[ih].seeders[preferredSubnet]; !ok {
```
Shoulde be `shard.swarms[ih].seeders[preferredSubnet][pk]`

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,780 @@
+			numWant--
+		}
+
+		// Append as many close leechers as possible.
```
Maybe wrap this whole block in `if numWant > 0` too

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,780 @@
+	default:
+	}
+
+	ihDelta := float64(0)
```
No need

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,780 @@
+	}
+
+	recordGCDuration(time.Since(start))
+	recordInfohashesDelta(ihDelta)
```
No need

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,780 @@
+
+	shard.numSeeders--
+	delete(shard.swarms[ih].seeders[preferredSubnet], pk)
+
```
Check if the subnet is empty and remove if yes?

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,780 @@
+
+	shard.numLeechers--
+	delete(shard.swarms[ih].leechers[preferredSubnet], pk)
+
```
Check for empty subnet and remove?

mrd0ll4r commented on storage/memorybysubnet/peer_store.go:
```diff
@@ -0,0 +1,768 @@
+
+	if shard.swarms[ih].lenSeeders()|shard.swarms[ih].lenLeechers() == 0 {
+		delete(shard.swarms, ih)
+	} else if shard.swarms[ih].lenSeeders() == 0 {
```
Oh has this been here all the time? Ignore my comment above then.
I think we should do `len(shard.swarms[ih].seeders[preferredSubnet] == 0` to clean up this subnet when it's empty. Otherwise we'd leave it as long as there is any subnet with at least one seeder left, doesn't feel right.

jzelinskie commented:
cc @cooperlees, you guys should be able to test against HEAD now

cooperlees commented:
I don't think so - We're all super forked now and using SSL on the same port as non SSL.
https://github.com/psaab/chihaya

jzelinskie commented:
Can you list the things we'd need to realign?

This is some of the stuff I can see at a glance:

* It'd be quite trivial to maintain a patch for cmux on top of HEAD and use that.
* The "graceful" library isn't necessary on modern versions of Go, so you don't need any of those patches
* It looks like you have a patch to reload TLS certs for HTTP. HEAD supports reloading config with a SIGUSR1 signal, maybe this is sufficient to cover that functionality?

cc @psaab

psaab commented:
We really don't want to reload the cert via a signal since we change certs
often. I know graceful isn't needed but had issues with cmux on shutdown so
I didn't spend any time on it
On Sun, Jun 4, 2017 at 2:14 PM Jimmy Zelinskie <notifications@github.com>
wrote:

> Can you list the things we'd need to realign?
>
> This is some of the stuff I can see at a glance:
>
>    - It'd be quite trivial to maintain a patch for cmux on top of HEAD
>    and use that.
>    - The "graceful" library isn't necessary on modern versions of Go, so
>    you don't need any of those patches
>    - It looks like you have a patch to reload TLS certs for HTTP. HEAD
>    supports reloading config with a SIGUSR1 signal, maybe this is sufficient
>    to cover that functionality?
>
> cc @psaab <https://github.com/psaab>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/chihaya/chihaya/pull/301#issuecomment-306067420>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAMBUmZ5lvUdSVrCfL4vTCWbI0tw_Fvzks5sAx4qgaJpZM4MG6UI>
> .
>
